### ISSUE TITLE
Different #kernels generated for fwd call with requires_grad=True/False

### ISSUE THREAD

**Issue by BoyuanFeng** [pytorch/pytorch#161326]
For the forward call of a function, running with `requires_grad=False` generates 3 kernels. But running with `requires_grad=True` generates 6 kernels. I assume this is due to saved tensor or graph partition. But saved tensors should be an extra output in the fwd call which does not need recomputation.

```
import torch
torch._inductor.config.force_disable_caches = True
# torch._inductor.config.triton.cooperative_reductions = True
@torch.compile(fullgraph=True, dynamic=False)
def _local_metrics(x):
    return x.abs().max(), x.abs().mean(), x.square().mean()
_local_metrics(torch.rand([512, 512, 512], device="cuda", requires_grad=True)) # <--- try switch between requires_grad=True/False
```

Slightly related issue: #161138

**Comment by BoyuanFeng** [comment]
The repro is actually tricky. For fwd function with `requires_grad=True`, we can observe 4 kernels in shape=[1024, 1024, 1024] and 6 kernels in shape=[512,512,512]. [P1914643122](https://www.internalfb.com/phabricator/paste/view/P1914643122)

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260516+cu132`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs observed the kernel-count asymmetry described in the issue. With shape=[512,512,512]: requires_grad=False -> generated_kernel_count=4, requires_grad=True -> generated_kernel_count=6. Counts were identical across all 3 runs and test_fn exited 1 each time.

<details>
<summary>Error output</summary>

```
requires_grad=False -> generated_kernel_count = 4
requires_grad=True  -> generated_kernel_count = 6
Mismatch: True (no_grad=4, with_grad=6)

/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_dynamo/pgo.py:596: UserWarning: dynamo_pgo force disabled by torch.compiler.config.force_disable_caches
  warn_once(
/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_dynamo/pgo.py:596: UserWarning: dynamo_pgo force disabled by torch.compiler.config.force_disable_caches
  warn_once(
/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_dynamo/pgo.py:596: UserWarning: dynamo_pgo force disabled by torch.compiler.config.force_disable_caches
  warn_once(
/home/karthickps/.conda/envs/pytorch-nightly/lib/python3.13/site-packages/torch/_dynamo/pgo.py:596: UserWarning: dynamo_pgo force disabled by torch.compiler.config.force_disable_caches
  warn_once(

ERROR conda.cli.main_run:execute(49): `conda run python /home/karthickps/.repro_agent/workdir/pytorch_pytorch/161326/repro.py` failed. (See above for error)
```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch
import torch._inductor.config
import torch._inductor.metrics

torch._inductor.config.force_disable_caches = True


def _local_metrics(x):
    return x.abs().max(), x.abs().mean(), x.square().mean()


def _count_kernels(requires_grad: bool, backend: str) -> int:
    torch._dynamo.reset()
    torch._inductor.metrics.reset()
    compiled = torch.compile(_local_metrics, fullgraph=True, dynamic=False, backend=backend)
    x = torch.rand([512, 512, 512], device="cuda", requires_grad=requires_grad)
    outs = compiled(x)
    for o in outs:
        _ = o.detach().cpu()
    return torch._inductor.metrics.generated_kernel_count


def test_fn() -> bool:
    """Return True if NO bug (kernel counts match). False if bug present."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
    n_no_grad = _count_kernels(requires_grad=False, backend=backend)
    n_with_grad = _count_kernels(requires_grad=True, backend=backend)
    # No bug if forward kernel counts are the same regardless of requires_grad.
    return n_no_grad == n_with_grad


if __name__ == "__main__":
    torch._dynamo.reset()
    torch._inductor.metrics.reset()
    compiled = torch.compile(_local_metrics, fullgraph=True, dynamic=False)
    x_ng = torch.rand([512, 512, 512], device="cuda", requires_grad=False)
    out_ng = compiled(x_ng)
    for o in out_ng:
        _ = o.detach().cpu()
    n_no_grad = torch._inductor.metrics.generated_kernel_count
    print(f"requires_grad=False -> generated_kernel_count = {n_no_grad}")

    torch._dynamo.reset()
    torch._inductor.metrics.reset()
    compiled = torch.compile(_local_metrics, fullgraph=True, dynamic=False)
    x_g = torch.rand([512, 512, 512], device="cuda", requires_grad=True)
    out_g = compiled(x_g)
    for o in out_g:
        _ = o.detach().cpu()
    n_with_grad = torch._inductor.metrics.generated_kernel_count
    print(f"requires_grad=True  -> generated_kernel_count = {n_with_grad}")

    print(f"Mismatch: {n_no_grad != n_with_grad} (no_grad={n_no_grad}, with_grad={n_with_grad})")

    try:
        ok = test_fn()
    except Exception:
        import traceback; traceback.print_exc()
        os._exit(2)
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260516+cu132
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.4.0.1
[pip3] nvidia-cuda-cupti==13.2.75
[pip3] nvidia-cuda-nvrtc==13.2.78
[pip3] nvidia-cuda-runtime==13.2.75
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.2.0.46
[pip3] nvidia-curand==10.4.2.55
[pip3] nvidia-cusolver==12.2.0.1
[pip3] nvidia-cusparse==12.7.10.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.2.78
[pip3] nvidia-nvtx==13.2.75
[pip3] torch==2.13.0.dev20260516+cu132
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260516+cu132
[pip3] torchdata==0.11.0
[pip3] torchmetrics==1.0.3
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260516+cu132
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
